
### recode the dense file for visualization

```bash
cd ~/scratch/TCGA_CCLE_SKCM/SKCM_MSTC_CCLE_RUN/snakemake_ChIPseq_pipeline/12chromHMM/16CCLE_10MSTC_20_SKCM_CHROMHMM_OUTPUT/dense_bed

# rename the dense file
brename -p "-P[0-9]*_18_dense" -r ""
ls

M028.bed  M137.bed  M263.bed  M305.bed  M357.bed  M409.bed  M642.bed  M749.bed  M807.bed  M852.bed  recode_dense_seg_state_color.py
M035.bed  M233.bed  M275.bed  M306.bed  M399.bed  M527.bed  M721.bed  M762.bed  M822.bed  M857.bed  my_map.txt

cat my_map.txt
old_state       new_state       new_color
1       1       0,0,255
2       1       0,0,255
3       2       0,153,204
4       3       0,255,255
5       3       0,255,255
6       4       51,255,0
7       5       0,204,51
8       6       0,102,0
9       7       0,0,51
10      8       204,0,51
11      9       255,0,204
12      10      204,153,255
13      11      204,204,102
14      12      255,255,0
15      13      255,255,204
16      2       0,153,204
17      14      51,102,102
18      15      102,153,51

## recode

find *bed | parallel './recode_dense_seg_state_color.py --ifile {} --file_type dense --map_file my_map.txt --ofile {/.}_recode.bed'
```


### heatmap after feature selection on tumor data.

The first step feature selection usually get the features down to ~3000 per chromosome, use chr7 for testing

On HPC.

```{bash}
cd /rsrch2/genomic_med/krai/scratch/SKCM_RoadTrip/pyflow-epilogos/09vsurf_output_by_chr
```

get the features

```{r}
library(tidyverse)
load("chr7_vsurf_out.rda")
ls()
#rf_mat<- test_data[vsurf$varselect.interp]
rf_mat<- x_data[,vsurf$varselect.thres]

write.table(rf_mat, "chr7_rf_mat.txt", row.names = T, col.names = T, sep = "\t", quote = F)

```

on local computer

```{bash}
cd  /Users/mtang1/github_repos/RoadTrip_epigenome/data/epigenome_SKCM/data
rsync -avhP railab:scratch/SKCM_RoadTrip/pyflow-epilogos/09vsurf_output_by_chr/chr7_rf_mat.txt .

```

```{r}
tumor_meta<- read_tsv("data/epigenome_SKCM/tumor_meta_data.txt", col_names =T)
tumor_meta<- mutate(tumor_meta, subtype = gsub("_Hotspot_Mutant", "", subtype))

rf_mat<- read.table("data/epigenome_SKCM/data/chr7_rf_mat.txt", header = T, row.names = 1)
# reorder the mat according to the meta data
rf_mat<- rf_mat[tumor_meta$sample, ]

rf_mat<- as.matrix(rf_mat)
rf_mat<- t(rf_mat)

## change to a factor, and set levels, this is important when convert to numeric and affect clustering

rf_numeric_mat<- rf_mat %>%
        as.data.frame() %>%
        mutate_all(factor, levels = c(paste0("E",c(1,3,6,7,8,9,4,5,13,14,15,12,11,2,10)))) %>%
        mutate_all(as.numeric) %>%
        as.matrix()

rownames(rf_numeric_mat)<- rownames(rf_mat)


```



```{r}
cell_fun = function(j, i, x, y, width, height, fill) {
	grid.rect(x = x, y = y, width = width *0.98, height = height *0.98,
		gp = gpar(col = "grey", fill = fill, lty = 1, lwd = 0.5))
}

## get colors http://wiki.wubrowse.org/QuantitativeCategorySeries
col_fun<- circlize::colorRamp2(1:15, c('#ffff00', '#c2e105', '#ff4500','#32cd32',
                                       '#ff0000','#e9967a',
                                       '#008000', '#006400',
                                       '#cd5c5c',
                                       '#8a91d0', '#66cdaa',
                                       '#c0c0c0', '#808080',
                                       '#ffffff', '#ffffff'))

df<- as.data.frame(tumor_meta)[,2, drop =F]
rownames(df)<- tumor_meta$sample

ha<-  HeatmapAnnotation(df = df, col = list(subtype = c("NRAS" =  "red", "BRAF" = "blue",
                                                       "Triple_WT" = "green")))
library(genefilter)
rv<- rowVars(rf_numeric_mat)
## select the top n most variable genes for clustering
idx<- order(-rv)[1:200]

idx<- apply(rf_numeric_mat, 1, function(x) sum(x==14)) >=5


Heatmap(rf_numeric_mat[-idx,], show_row_names = F,
        col = col_fun,
        cluster_columns = T,
        row_dend_reorder = T,
        column_dend_reorder = T,
        heatmap_legend_param = list(color_bar = "discrete", at = 1:15,
                                    labels = c('Active Enhancer', 'Transcribed Enhancer',
                                               'Transcribed tss K27ac low',
                                               'Transcribed tss K27ac high',
                                               'Active tss K4me1 high', 'Active tss K4me1 low',
                                               'Strong transcription', 'Weak transcription',
                                               'trivalent tss', 'Weak Heterochromatin',
                                               'Strong Heterochromatin','Weak PcG repressive',
                                               'Strong PcG repressive', 'Low state',
                                               'Artifact')),
        name = "chromatin state",
        top_annotation = ha)

```


```{bash}
1       E1      Active Enhancer
2       E3      Transcribed Enhancer
3       E6      Transcribed tss K27ac low
4       E7      Transcribed tss K27ac high
5       E8      Active tss K4me1 high
6       E9      Active tss K4me1 low
7       E4      Strong transcription
8       E5      Weak transcription
9       E13     trivalent tss
10      E14     Weak Heterochromatin
11      E15     Strong Heterochromatin
12      E12     Weak PcG repressive
13      E11     Strong PcG repressive
14      E2      Low state
15      E10     Artifact


```
